File: 143.重排链表/reorderList.go
package problem0143

type ListNode struct {
	Val  int
	Next *ListNode
}

// 找中点（右中点）= 876题
func middleNode(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

// 翻转链表 = 206题
func reverseList(head *ListNode) *ListNode {
	var pre *ListNode = nil
	cur := head
	for cur != nil {
		next := cur.Next
		cur.Next = pre
		pre, cur = cur, next
	}
	return pre
}

// 从中点分隔链表，原地反转后半部分，迭代这两个链表得到答案 = 876题 + 206题
func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}
	mid := middleNode(head)
	left := head
	right := reverseList(mid.Next)
	mid.Next = nil

	for right != nil {
		lNext := left.Next
		rNext := right.Next
		left.Next = right
		right.Next = lNext
		left = lNext
		right = rNext
	}
}

File: 143.重排链表/reorderList.js
// 重排链表
// 将链表 a -> b -> c -> d -> e
// 转变为 a -> e -> b -> d -> c

function getMiddle(head) {
	let slow = head
	let fast = head
	while (fast !== null && fast.next !== null) {
		slow = slow.next
		fast = fast.next.next
	}
	return slow
}

function reverse(head) {
	let prev = null
	let cur = head
	while (cur !== null) {
		const next = cur.next
		cur.next = prev
		prev = cur
		cur = next
	}
	return prev
}

function mergeList(a, b) {
	let aCache = null
	let bCache = null
	while (a !== null && b !== null) {
		aCache = a.next
		bCache = b.next
		a.next = b
		a = aCache
		b.next = a
		b = bCache
	}
}

function reorderList(head) {
	if (!head || !head.next) {
		return head
	}
	// 先找中点
	const mid = getMiddle(head)
	// 再反转中点后面的节点
	const right = reverse(mid.next)
	// 斩断中点后面的引用
	mid.next = null
	mergeList(head, right)
}